src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/GCHandle.cs

```diff
@@ -67,13 +66,34 @@ private GCHandle(object? value, GCHandleType type)
         /// <returns>A new GC handle that protects the object.</returns>
         public static GCHandle Alloc(object? value, GCHandleType type) => new GCHandle(value, type);
 
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            IntPtr handle = _handle;
+
+            if ((nint)handle != 0)
+            {
+                _handle = IntPtr.Zero;
+
+                InternalFree(GetHandleValue(handle));
+            }
+        }
+
         /// <summary>Frees a GC handle.</summary>
         public void Free()
         {
-            // Free the handle if it hasn't already been freed.
-            IntPtr handle = Interlocked.Exchange(ref _handle, IntPtr.Zero);
-            ThrowIfInvalid(handle);
-            InternalFree(GetHandleValue(handle));
+            IntPtr handle = _handle;
+
+            if ((nint)handle != 0)
+            {
+                _handle = IntPtr.Zero;
+
+                InternalFree(GetHandleValue(handle));
+            }
+            else
+            {
+                ThrowHelper.ThrowInvalidOperationException_HandleIsNotInitialized();
+            }
```

Comment:
Why not use ThrowIfInvalid?

Reply:
```C#
IntPtr handle = _handle;
ThrowIfInvalid(handle);
_handle = IntPtr.Zero;
InternalFree(GetHandleValue(handle));
```

Reply:
`ThrowIfInvalid` takes the input handle and checks it for `null` again, which is not needed here since this branch will always throw. I guess the JIT might likely be able to optimize that away but I figured I'd just save it some trouble and also make the IL a tiny bit smaller due to no `ldloc.x` needed to pass the field value? 🤔

Reply:
> checks it for null again

What do you mean by "again"?  It'd be the only check.

Reply:
What I meant is that right now we have this:

https://github.com/dotnet/runtime/blob/ec81ed110f3173edb87b5316e1bd9b99c111e725/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/GCHandle.cs#L87-L96

So we're already checking for `null` on that local handle here. I just saw your other comment after your first one and I now understand that you were talking about removing the conditional branch inline in the `Free` method entirely and just unconditionally call `ThrowIfInvalid` first. I can do that, sure! Mostly just had done this for consistency with the code in `DependentHandle` I guess, but I'm happy to switch this to what you suggested 😄